provisioner/terraform/otelenv.go

```diff
@@ -0,0 +1,74 @@
+package terraform
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"unicode"
+
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/propagation"
+)
+
+// TODO: replace this with the upstream OTEL env propagation when it is
+// released.
+
+// envCarrier is a propagation.TextMapCarrier that is used to extract or
+// inject tracing environment variables. This is used with a
+// propagation.TextMapPropagator
+type envCarrier struct {
+	Env []string
+}
+
+var _ propagation.TextMapCarrier = (*envCarrier)(nil)
+
+func toKey(key string) string {
+	key = strings.ToUpper(key)
+	key = strings.ReplaceAll(key, "-", "_")
+	return strings.Map(func(r rune) rune {
+		if unicode.IsLetter(r) || unicode.IsNumber(r) || r == '_' {
+			return r
+		}
+		return -1
+	}, key)
+}
+
+func (c *envCarrier) Set(key, value string) {
+	if c == nil {
+		return
+	}
+	key = toKey(key)
+	for i, e := range c.Env {
+		if strings.HasPrefix(e, key+"=") {
+			// don't directly update the slice so we don't modify the slice
+			// passed in
+			newEnv := slices.Clone(c.Env)
+			newEnv = append(newEnv[:i], append([]string{fmt.Sprintf("%s=%s", key, value)}, newEnv[i+1:]...)...)
```

Comment:
Once you've cloned the initial slice, seems like it would be simpler to just modify in place, rather than re-slice and append.

Reply:
hah, I think that code went through too many refactors and came out nonsense. I have simplified it.